src/main/java/omero/cmd/graphs/DuplicateI.java

```diff
@@ -137,13 +179,98 @@ public DuplicateI(ACLVoter aclVoter, Roles securityRoles, GraphPathBean graphPat
+        this.managedRepository = applicationContext.getBean(ManagedRepositoryI.class);
+        this.pathServices = ImmutableMap.of(
+                OriginalFile.class, applicationContext.getBean("/OMERO/Files", OriginalFilesService.class)::getFilesPath,
+                Pixels.class, applicationContext.getBean("/OMERO/Pixels", PixelsService.class)::getPixelsPath,
+                Thumbnail.class, applicationContext.getBean("/OMERO/Thumbs", ThumbnailService.class)::getThumbnailPath);
     }
 
     @Override
     public Map<String, String> getCallContext() {
         return null;
     }
 
+    /**
+     * Create a duplicate of a file on the filesystem.
+     * @param from the file to duplicate
+     * @param to where to create the duplicate
+     * @throws IOException if the file duplication failed
+     */
+    private void duplicateFile(final Path from, final Path to) throws IOException {
+        final boolean isFromManaged = isPathManaged.test(from);
+        final boolean isToManaged = isPathManaged.test(to);
+        if (isFromManaged && isToManaged) {
+            try {
+                /* safe to create a hard link in managed repository */
+                Files.createLink(to, from);
+                LOGGER.debug("linked {} to {}", from, to);
+            } catch (FileSystemException fse) {
+                LOGGER.debug("failed to link {} to {}", from, to, fse);
+                /* cannot link so fall back to a copy */
+                Files.copy(from, to);
+                LOGGER.debug("copied {} to {}", from, to);
+            }
+        } else if (isFromManaged || isToManaged) {
+            throw new IllegalArgumentException("either both or neither of the files may be in the managed repository");
+        } else {
+            /* copy rather than link as may violate integrity if different users can write to same file */
+            Files.copy(from, to);
+            LOGGER.debug("copied {} to {}", from, to);
+        }
+    }
+
+    /**
+     * Get the underlying filesystem path for the given object.
+     * @param modelObject a model object
+     * @return the underlying filesystem path for the object, or {@code null} if none exists
+     * @throws GraphException if the object should not be included in duplication
+     */
+    private Path getPath(IObject modelObject) throws GraphException {
+        if (modelObject instanceof OriginalFile) {
+            /* Files may require special handling. */
+            final OriginalFile file = (OriginalFile) modelObject;
+            String fileRepo = file.getRepo();
+            if (fileRepo == null) {
+                fileRepo = originalFileRepos.get(file.getId());
+            }
+            if (PublicRepositoryI.DIRECTORY_MIMETYPE.equals(file.getMimetype())) {
+                /* Do not attempt to duplicate directories. */
+                LOGGER.debug("not directly duplicating OriginalFile:{} on filesystem because it is a directory", file.getId());
```

Comment:
What's the impact of not duplicating the directories?

Reply:
Simplifies the code and adapts existing code. :smiley: We can focus on just translating filenames and creating their parents without also separately having to recognize and translate partial paths to their new location.